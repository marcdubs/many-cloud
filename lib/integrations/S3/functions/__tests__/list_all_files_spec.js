describe("list_all_files", () => {
  describe("when not provided a folderID", () => {
    const expected_marker = "WASUP";
    let data, actual;
    beforeEach(async () => {
      data = {
        s3: {
          listObjects: jest.fn().mockImplementation((params, callback) => {
            if (params.Marker === expected_marker) {
              callback(null, {
                IsTruncated: false,
                Contents: ["Thing2"],
                CommonPrefixes: []
              });
            } else {
              callback(null, {
                IsTruncated: true,
                NextMarker: expected_marker,
                Contents: ["Thing1"],
                CommonPrefixes: []
              });
            }
          })
        }
      };
      actual = await require("../list_all_files")(data)(null);
    });

    it("calls data.s3.listObjects twice", () => {
      expect(data.s3.listObjects).toHaveBeenCalledTimes(2);
    });

    it("returns the expected contents", () => {
      expect(actual.Contents).toMatchSnapshot();
    });
  });

  describe("when provided a folderID", () => {
    const expected_marker = "WASUP";
    let data, actual;
    beforeEach(async () => {
      data = {
        s3: {
          listObjects: jest.fn().mockImplementation((params, callback) => {
            if (params.Marker === expected_marker) {
              callback(null, {
                IsTruncated: false,
                Contents: ["Thing2"],
                CommonPrefixes: ["SomeFolder"]
              });
            } else {
              callback(null, {
                IsTruncated: true,
                NextMarker: expected_marker,
                Contents: ["Thing1"],
                CommonPrefixes: []
              });
            }
          })
        }
      };
      actual = await require("../list_all_files")(data)("folder");
    });

    it("calls data.s3.listObjects twice", () => {
      expect(data.s3.listObjects).toHaveBeenCalledTimes(2);
    });

    it("calls data.s3.listObjects with the folder as the key prefix", () => {
      expect(data.s3.listObjects.mock.calls).toMatchSnapshot();
    });

    it("returns the expected contents", () => {
      expect(actual.Contents).toMatchSnapshot();
    });
  });

  describe("when listObject throws an error the second time it is called", () => {
    const expected_marker = "WASUP";
    const expected_error = "OH BOTHER!";
    let data, actual;
    beforeEach(() => {
      data = {
        s3: {
          listObjects: jest.fn().mockImplementation((params, callback) => {
            if (params.Marker === expected_marker) {
              callback(new Error(expected_error), null);
            } else {
              callback(null, {
                IsTruncated: true,
                NextMarker: expected_marker,
                Contents: ["Thing1"],
                CommonPrefixes: []
              });
            }
          })
        }
      };
    });

    it("returns with the error", async () => {
      let func = require("../list_all_files")(data);
      await expect(func(null)).rejects.toThrowErrorMatchingSnapshot();
    });
  });
});
