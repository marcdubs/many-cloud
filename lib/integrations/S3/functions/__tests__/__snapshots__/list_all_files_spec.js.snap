// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`list_all_files when listObject throws an error the second time it is called returns with the error 1`] = `"OH BOTHER!"`;

exports[`list_all_files when not provided a folderID returns the expected contents 1`] = `
Array [
  "Thing1",
  "Thing2",
]
`;

exports[`list_all_files when provided a folderID calls data.s3.listObjects with the folder as the key prefix 1`] = `
Array [
  Array [
    Object {
      "Bucket": undefined,
      "Delimiter": "/",
      "Marker": "WASUP",
      "Prefix": "folder/",
    },
    [Function],
  ],
  Array [
    Object {
      "Bucket": undefined,
      "Delimiter": "/",
      "Marker": "WASUP",
      "Prefix": "folder/",
    },
    [Function],
  ],
]
`;

exports[`list_all_files when provided a folderID returns the expected contents 1`] = `
Array [
  "Thing1",
  "Thing2",
]
`;
